index.js

```diff
@@ -30,26 +43,23 @@ class BindingComponent extends React.Component {
         val = this.props.bindStateType(val);
       }
     }
-    const names = this.props.bindStateName.split('.');
-    if (names.length === 1) {
-      this.props.bindStateCtx.setState({
-        [names[0]]: val
-      });
-    } else {
-      const newState = this.props.bindStateCtx.state[names[0]];
-      const subnames = names.slice(1);
-      subnames.reduce((curr, name, index) => {
-        if (index === subnames.length - 1) {
-          curr[name] = val;
-        } else {
-          curr = curr[name];
-        }
-        return curr;
-      }, newState);
-      this.props.bindStateCtx.setState({
-        [names[0]]: newState
-      });
-    }
+
+    this.props.bindStateCtx.setState((prevState) => {
+      let pathReversed;
+      let overlay = val;
+      let pathIndex = 0;
+      try {
+        pathReversed = this.props.bindStateName.split('.').reverse();
+      } catch (error) {
+        return prevState;
+      }
+      do {
+        let subpath = pathReversed[pathIndex];
+        overlay = { [subpath]: overlay };
+        pathIndex++;
+      } while (pathIndex < pathReversed.length);
+      return assign(prevState, overlay);
```

Comment:
@scottoasis The first argument as function type of the `setState` should return the diff layer, not the complete state object although that is working. This is the core reason that I suggest to avoid calling *deep-assign*.